pyo3-ffi/README.md

```diff
@@ -0,0 +1,188 @@
+#
+# Downstream Rust code (including code in `bin/`, `examples/`, and `tests/`) will not be able
+# to `use string_sum;` unless the "rlib" or "lib" crate type is also included, e.g.:
+# crate-type = ["cdylib", "rlib"]
+crate-type = ["cdylib"]
+
+[dependencies.pyo3-ffi]
+orkaround for `extended_key_value_attributes`: https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+fg_attr(docsrs, cfg_attr(docsrs, doc = concat!("version = \"", env!("cargo_pkg_version"),  "\"")))]
+fg_attr(not(docsrs), doc = "version = \"*\"")]
+features = ["extension-module"]
+```
+
+**`src/lib.rs`**
+```rust
+use std::intrinsics::transmute;
+use std::os::raw::c_char;
+
+use pyo3_ffi::*;
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn PyInit_string_sum() -> *mut PyObject {
+    let init = PyModuleDef {
+        m_base: PyModuleDef_HEAD_INIT,
+        m_name: "string_sum\0".as_ptr() as *const c_char,
+        m_doc: std::ptr::null(),
+        m_size: 0,
+        m_methods: std::ptr::null_mut(),
+        m_slots: std::ptr::null_mut(),
+        m_traverse: None,
+        m_clear: None,
+        m_free: None,
+    };
+
+    let mptr = PyModule_Create(Box::into_raw(Box::new(init)));
+    let version = env!("CARGO_PKG_VERSION");
+    PyModule_AddObject(
+        mptr,
+        "__version__\0".as_ptr() as *const c_char,
+        PyUnicode_FromStringAndSize(version.as_ptr() as *const c_char, version.len() as isize),
+    );
+
+    let wrapped_sum_as_string = PyMethodDef {
+        ml_name: "sum_as_string\0".as_ptr() as *const c_char,
+        ml_meth: Some(transmute::<_PyCFunctionFast, PyCFunction>(sum_as_string)),
+        ml_flags: METH_FASTCALL,
+        ml_doc: "returns the sum of two integers as a string".as_ptr() as *const c_char,
+    };
+
+    PyModule_AddObject(
+        mptr,
+        "sum_as_string\0".as_ptr() as *const c_char,
+        PyCFunction_NewEx(
+            Box::into_raw(Box::new(wrapped_sum_as_string)),
+            std::ptr::null_mut(),
+            PyUnicode_InternFromString("string_sum\0".as_ptr() as *const c_char),
+        ),
+    );
+
+    let all = ["__all__\0", "__version__\0", "sum_as_string\0"];
+
+    let pyall = PyTuple_New(all.len() as isize);
+    for (i, obj) in all.iter().enumerate() {
+        PyTuple_SET_ITEM(
+            pyall,
+            i as isize,
+            PyUnicode_InternFromString(obj.as_ptr() as *const c_char),
+        )
+    }
+
+    PyModule_AddObject(mptr, "__all__\0".as_ptr() as *const c_char, pyall);
+
+    mptr
+}
+
+pub unsafe extern "C" fn sum_as_string(
+    _self: *mut PyObject,
+    args: *mut *mut PyObject,
+    nargs: Py_ssize_t,
+) -> *mut PyObject {
+    if nargs != 2 {
+        return raise_type_error("sum_as_string() expected 2 positional arguments");
+    }
+
+    let arg1 = *args;
+    if PyLong_Check(arg1) == 0 {
+        return raise_type_error("sum_as_string() expected an int for positional argument 1");
+    }
+    let arg1 = PyLong_AsLong(arg1);
```

Comment:
`PyLong_AsLong` will return an error on overflow, which must be handled (same for the other one)

Reply:
Good catch. Will fix this. Thanks :+1: